### DISCUSSION TITLE
Help with CSRF on Rodauth + Omniauth - API Only (Secure cookie)

### DISCUSSION THREAD

**Discussion by AlexeyMatskevich** [janko/rodauth-omniauth#23] [Q&A] [answered]
Greetings! I've set up a new application using the latest versions of everything. In the context of my question, I've configured Rails, Rodauth-rails, and Rodauth-omniauth. (I once implemented Argon2 integration in Rodauth and migration mechanisms and was pleasantly surprised by the integration in Rodauth-rails—thanks to the author!)

In our application, we don’t plan to use multiple domains, so we intend to use cookies for user identification. As I understand, this is well-supported by Omniauth itself and, seemingly, also by Rodauth and Rodauth-rails, though I might be mistaken. Ideally, I would like to enable CSRF protection to generally reduce the potential risks associated with using cookies.

I’ve seen there is a guide for using Omniauth in JSON (i.e., API) mode and saw mentions of built-in CSRF protection without using an external gem. However, I couldn’t find any information in the documentation or discussions on how to use CSRF protection in API mode.

In a discussion similar to this one [link](https://github.com/janko/rodauth-omniauth/discussions/17), I encountered the following error:

```shell
Started GET "/auth/github/callback?code=...&state=..." for 127.0.0.1 at 2024-10-31 23:28:28 +0500
  ActiveRecord::SchemaMigration Load (1.2ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC /*application='RailsBackend'*/
D, [2024-10-31T23:28:28.653328 #3049637] DEBUG -- omniauth: (github) Setup endpoint detected, running now.
D, [2024-10-31T23:28:28.653372 #3049637] DEBUG -- omniauth: (github) Callback phase initiated.
E, [2024-10-31T23:28:28.656543 #3049637] ERROR -- omniauth: (github) Authentication failure! csrf_detected: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected
Started GET "/favicon.ico" for 127.0.0.1 at 2024-10-31 23:28:28 +0500
```
At this stage, I don’t understand how to correctly enable CSRF protection and assume it should be implemented in the Rails way through X-CSRF-Token.

Moreover, I believe there are two distinct phases for CSRF tokens:

1. The authorization session—POST /auth/github request and the callback GET /auth/github/callback?code=...&state=... where, as I understand it, a CSRF token is needed to avoid the vulnerability described here, although I don’t fully understand the nature of this vulnerability to know why exactly a CSRF token is needed in this context.

2. The issuance of a CSRF token after authorization.

I will elaborate on the second phase. When the backend is a public API, as I understand it, it’s not enough to simply issue a CSRF token since it could easily be obtained, rendering it ineffective. Therefore, it would likely be correct to issue the CSRF token only during authentication.

Currently, I'm not entirely sure how this should work, and I couldn't immediately find any guidance or recommendations in the Rodauth/Rodauth-rails documentation—on how to issue a CSRF token in API mode or in similar edge cases. I believe this question likely relates directly to Rodauth itself, stemming from its initial design for server-side rendering, with API mode support provided by the community.

**Comment by AlexeyMatskevich** [comment]
Strangely, this solution https://github.com/janko/rodauth-omniauth/discussions/17#discussioncomment-8496431  doesn't allow me to get rid of the error, I was planning to use this at least as a temporary measure.
```shell
Started POST "/auth/github" for 127.0.0.1 at 2024-11-01 01:07:58 +0500
D, [2024-11-01T01:07:58.348066 #3085463] DEBUG -- omniauth: (github) Setup endpoint detected, running now.
D, [2024-11-01T01:07:58.348131 #3085463] DEBUG -- omniauth: (github) Request phase initiated.
Started GET "/auth/github/callback?code=...&state=..." for 127.0.0.1 at 2024-11-01 01:08:05 +0500
D, [2024-11-01T01:08:05.620654 #3085463] DEBUG -- omniauth: (github) Setup endpoint detected, running now.
D, [2024-11-01T01:08:05.620679 #3085463] DEBUG -- omniauth: (github) Callback phase initiated.
E, [2024-11-01T01:08:05.622222 #3085463] ERROR -- omniauth: (github) Authentication failure! csrf_detected: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected
Started GET "/favicon.ico" for 127.0.0.1 at 2024-11-01 01:08:05 +0500
```

**Comment by janko** [comment]
What method of handling unverified requests are you using in `protect_from_forgery`? If it's `:null_session`, that would explain the behavior you're seeing:

1. CSRF token wasn't valid in the request/authorize phase, and Rails continues the request, but uses a null session
1. omniauth-oauth2 attempts to store the "state" (random string) into the session, but that doesn't get persisted
1. in the callback phase, the state is read from the URL and compared to the state from session, which is missing
1. omniauth-oauth2 concludes that the mismatch means it's a CSRF issue and fails

In that case, you have to figure out why the CSRF check failed for the request phase. Rails should read it from `X-CSRF-Token` header, which means rodauth-omniauth will as well in combination with rodauth-rails. Any chance the `Origin` header check failed, which is also part of Rails' CSRF protection? Do you see anything in the Rails logs for the request phase?

**Reply by janko** [reply]
Note that Rails won't check the CSRF token in the callback phase, because that's a GET request, and Rails skips CSRF checks for GET requests.

> Currently, I'm not entirely sure how this should work, and I couldn't immediately find any guidance or recommendations in the Rodauth/Rodauth-rails documentation—on how to issue a CSRF token in API mode or in similar edge cases. 

The CSRF token for rodauth-omniauth is hopefully issued in the same way you'd issue a CSRF token for any other Rails API endpoint, that's why it's not documented.

> I believe this question likely relates directly to Rodauth itself, stemming from its initial design for server-side rendering, with API mode support provided by the community.

Unlike Devise, whose API mode is provided by 3rd-party extensions, Rodauth's JSON API mode is built-into the framework, so it's native and maintained together with the rest of Rodauth.

**Reply by AlexeyMatskevich** [reply]
> Unlike Devise, whose API mode is provided by 3rd-party extensions, Rodauth's JSON API mode is built-into the framework, so it's native and maintained together with the rest of Rodauth.

I just mean that, based on my four years of experience with Rodauth, everything works fine in production when using server-side rendering. However, when using the API, I’ve always had to contribute fixes or override or use hooks, as it's clear that Jeremy (the author of Rodauth) doesn’t use this mode himself; this mode has mostly been developed by other contributors, particularly by you. I don’t mean this as a criticism—it’s just a feature I take into account.

For example, in API Mode, there’s no documentation for any feature in Rodauth. To understand what data needs to be sent and what response to expect, one has to refer to the API mode tests. But I can’t send a frontend developer to look at tests written in Ruby. In this regard, I could contribute by adding an OpenAPI generator like [rspec-openapi](https://github.com/exoego/rspec-openapi) (where I’ve already added support for Rack and Hanami), which could be used to automatically generate OpenAPI specifications for all features based on tests, though it’s not a perfect tool.

Other examples that made me feel like no one had used Rodauth for API mode in production before me are my list of PRs, which I submitted when I integrated Rodauth for the first time in 2020:
    [#110](https://github.com/jeremyevans/rodauth/pull/110): The documentation described behavior that wasn’t implemented in the code for JWT refresh.
    [#133](https://github.com/jeremyevans/rodauth/pull/133): Ensuring JWT refresh behavior conforms to RFC 6750.
    [#148](https://github.com/jeremyevans/rodauth/pull/148): No need to check the password when creating an account (this only caused issues in the API, but I honestly don’t remember the details well anymore).
    
Anyway my opinion on this, though, is just my subjective opinion.

> What method of handling unverified requests are you using in protect_from_forgery?
No, I haven't configured protect_from_forgery, and in fact, I discovered that it was a naive assumption on my part that, because the error reported a CSRF issue, this meant that CSRF protection had automatically been enabled—but it appears that this was fundamentally incorrect.

It seems that, in this case, CSRF protection involves issuing a cookie with a token at the `POST /auth/:provider_name` stage, expecting that this token will automatically be returned with the cookie at the callback stage.

In my case, the problem turned out to be that I was using Firefox DevTools to send the POST request. After spending a lot of time, I realized that the Set-Cookie headers sent this way were ignored, and I couldn’t find settings like credentials: include ([MDN reference](https://developer.mozilla.org/en-US/docs/Web/API/Request/credentials)) in the DevTools UI.

Thus, the entire problem leading to the error Authentication failure! csrf_detected: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected was due to the absence of cookies. As soon as I sent the request via fetch with the credentials: "include" option, everything worked. Perhaps it would have helped if the cookie requirement was explicitly mentioned in the JSON flow, but the open question remains: should this be addressed in Omniauth or here in Rodauth-omniauth?

> The CSRF token for rodauth-omniauth is hopefully issued in the same way you'd issue a CSRF token for any other Rails API endpoint, that's why it's not documented.

Do I understand correctly that the developer determines how to initially provide the CSRF token to the client?

**Reply by janko** [reply]
> However, when using the API, I’ve always had to contribute fixes or override or use hooks, as it's clear that Jeremy (the author of Rodauth) doesn’t use this mode himself; this mode has mostly been developed by other contributors, particularly by you.

Yeah, that's true. JSON API mode (+ JWT) was added as a community request, I also doubt Jeremy himself uses it. I agree the JSON API path could be more polished, thanks for contributing your fixes along the way 👍🏻 

> For example, in API Mode, there’s no documentation for any feature in Rodauth.

Not sure if you caught it, but I actually released [rodauth-openapi](https://github.com/janko/rodauth-openapi) just recently, which aims to do precisely what you described 😉 Some of the things it handles:

* adapts to `*_route` and `*_param` changes (it even excludes routes that were disabled)
* changes whether a param is required based on configuration
* handles both HTML and JSON mode
* shows all possible error responses

> Perhaps it would have helped if the cookie requirement was explicitly mentioned in the JSON flow, but the open question remains: should this be addressed in Omniauth or here in Rodauth-omniauth?

Yeah, good point, I'll mention it in the rodauth-omniauth docs 👍🏻 It would be good if the error was more explicit, because "CSRF detected" doesn't tell you that cookies weren't persisted. That's probably best solved in omniauth-oauth2, since not all OmniAuth strategies are OAuth-based, and I would like to avoid doing any guesswork in rodauth-omniauth.

**Reply by AlexeyMatskevich** [reply]
> Not sure if you caught it, but I actually released [rodauth-openapi](https://github.com/janko/rodauth-openapi) just recently

I absolutely missed this, it's incredible!

**Reply by janko** [reply]
It's worth noting that I haven't yet updated rodauth-omniauth to utilize it, that's in my TODO list 🙂 But all the features that ship with Rodauth features are supported.

**Reply by AlexeyMatskevich** [reply]
I any recomendation about ?

> Do I understand correctly that the developer determines how to initially provide the CSRF token to the client?

**Reply by janko** [reply]
Right, yes, that sounds correct 👍🏻  I don't have much experience with a SPA + API architecture, so I don't know how the CSRF token is typically given to the client in this scenario.

**Reply by AlexeyMatskevich** [reply]
I plan to issue a token after successful authorization, I think I saw in discussions how this is done.
That is, in my architecture it will not be possible to request a token again without re-authorization. I think that in the case when the  token is available simply via API, this will make a CSRF request more possible for attackers.

**Reply by janko** [reply]
My recent research led me to conclude that JSON APIs aren't actually vulnerable to CSRF attacks, so the most recent version of Rodauth now disables CSRF protection for JSON APIs by default. In this case, you don't need to deal with CSRF tokens 😉

**Comment by AlexeyMatskevich** [comment, marked as answer]
Lack of a cookie received at the `POST /auth/:provider_name` step on the client results in `Authentication Error! csrf_detected: OmniAuth::Strategies::OAuth2::CallbackError, csrf_detected | CSRF detected ` at the `callback` step when your strategy is based on `omniauth-oauth2`

**Reply by benoror** [reply]
Related discussion: https://github.com/janko/rodauth-omniauth/discussions/17#discussioncomment-13139323
